core/src/apps/cardano/sign_tx/signer.py

```diff
@@ -379,6 +440,95 @@ def _is_change_output(self, output: messages.CardanoTxOutput) -> bool:
         """Used only to determine what message to show to the user when confirming sending."""
         return output.address_parameters is not None
 
+    async def _process_legacy_output(
+        self,
+        output_list: HashBuilderList,
+        output: messages.CardanoTxOutput,
+        should_show: bool,
+    ) -> None:
+        address = self._get_output_address(output)
+        output_list.append(address)
+
+        if output.asset_groups_count == 0:
+            # Output structure is: [address, amount, datum_hash?]
+            output_list.append(output.amount)
+        else:
+            # Output structure is: [address, [amount, asset_groups], datum_hash?]
+            output_value_list: HashBuilderList = HashBuilderList(2)
+            with output_list.append(output_value_list):
+                await self._process_output_value(output_value_list, output, should_show)
+
+        if output.datum_hash is not None:
+            if should_show:
+                await layout.confirm_datum_hash(self.ctx, output.datum_hash)
+            output_list.append(output.datum_hash)
+
+    async def _process_post_alonzo_output(
+        self,
+        output_dict: HashBuilderDict[int, Any],
+        output: messages.CardanoTxOutput,
+        should_show: bool,
+    ) -> None:
+        address = self._get_output_address(output)
+        output_dict.add(POST_ALONZO_OUTPUT_KEY_ADDRESS, address)
+
+        if output.asset_groups_count == 0:
+            # Only amount is added to the dict.
+            output_dict.add(POST_ALONZO_OUTPUT_KEY_AMOUNT, output.amount)
+        else:
+            # [amount, asset_groups] is added to the dict.
+            output_value_list: HashBuilderList = HashBuilderList(2)
+            with output_dict.add(POST_ALONZO_OUTPUT_KEY_AMOUNT, output_value_list):
+                await self._process_output_value(output_value_list, output, should_show)
+
+        if output.datum_hash is not None:
+            if should_show:
+                await layout.confirm_datum_hash(self.ctx, output.datum_hash)
+            output_dict.add(POST_ALONZO_OUTPUT_KEY_DATUM, output.datum_hash)
```

Comment:
Right, thanks for noticing!

Reply:
Hm, but simply wappring inline datum like `[1, inline_datum_cbor]` won't work, because `inline_datum_cbor` is a hash builder, not a primitive type. I think we need a separate `HashBuilderList` for this.